Listviews update

ErikKBethke opened the pull request:
Update list view functions for layers and documents. Changes made were to include current date in filename through datetime, adding headers to file, and additional fields.

Would appreciate confirmation on whether filename will work in proposed format and if all field names are correct.

ErikKBethke committed:
Add headers, CSV name, fields to document_list

Needs confirmation on API for field names. Also not sure if it is acceptable to import datetime and if it is possible to use this in filename creation.

ErikKBethke committed:
Add headers, CSV name, fields to layer_list

Needs confirmation on API for field names. Also not sure if it is acceptable to import datetime and if it is possible to use this in filename creation.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -489,18 +490,57 @@ def document_metadata_detail(request, docid, template='documents/document_metada
 @login_required
 def document_list(request):
     documents = Document.objects.all()
+	filename = (str(datetime.datetime.now()) + "_documents.csv")
```
Can you use spaces for indentation instead of tabs ?

ErikKBethke replied:
Tabs were replaced with spaces

Vampouille commented on geonode/documents/views.py:
```diff
@@ -489,18 +490,57 @@ def document_metadata_detail(request, docid, template='documents/document_metada
 @login_required
 def document_list(request):
     documents = Document.objects.all()
+	filename = (str(datetime.datetime.now()) + "_documents.csv")
```
I think you can also remove parenthesis around string definition : `filename = str(datetime.datetime.now()) + "_documents.csv"` should work

ErikKBethke replied:
Thank you, removed superfluous parentheses.

Vampouille commented on geonode/layers/views.py:
```diff
@@ -508,18 +509,67 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
 @login_required
 def layer_list(request):
     layer = Layer.objects.all()
+	filename = (str(datetime.datetime.now()) + "_layers.csv")
```
Here, you can remove surrounded parenthesis.

Vampouille commented on geonode/layers/views.py:
```diff
@@ -508,18 +509,67 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
 @login_required
 def layer_list(request):
     layer = Layer.objects.all()
```
`layer` var is not used so you can remove it

Vampouille commented on geonode/layers/views.py:
```diff
@@ -508,18 +509,67 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
+	filename = (str(datetime.datetime.now()) + "_layers.csv")
     response = HttpResponse(content_type='text/csv')
-    response['Content-Disposition'] = 'attachment; filename="geonode.csv"'
+    # response['Content-Disposition'] = 'attachment; filename="geonode_layers.csv"'  #ALTERNATIVE NAME IF DATETIME DOESNT WORK
```
I think you need a 'Content-Disposition' header for your response. Can you explain your comment : "If datetime doesn't work" ?

ErikKBethke replied:
This was an extra line that should not have made it into the update, thank you for pointing it out.

Vampouille reviewed (changes requested):
Can you indent with spaces instead of tab. I also made some other comments on code.

ErikKBethke committed:
Update views after review

User @Vampouille provided review that was taken into consideration and changes were made.

ErikKBethke committed:
Update views after review

User @Vampouille provided review that was taken into consideration and changes were made.

ErikKBethke commented:
Changes were all made, looking for re-review to merge request.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -502,5 +524,23 @@ def document_list(request):
+                         document.category,
+                         document.abstract
+                         ])
+return response
```
I think you need to indent this block to include it in previous function `document_list`

ErikKBethke replied:
Proper indentation added.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -489,9 +490,30 @@ def document_metadata_detail(request, docid, template='documents/document_metada
+                    'Abstract'
+                    ])
     for document in documents:
         t = filter(lambda x: x in string.printable, document.title)
```
Can you rename `t` variable to `title` and `i` to `id`, this will increase readability.

ErikKBethke replied:
Changed variable names to title, docid. I believe docid is more consitently used throughout the script, and id would not function properly.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -502,5 +524,23 @@ def document_list(request):
-    return response
+        writer.writerow([i,
+                         t,
+                         document.creation_date,
```
Missing document name value

ErikKBethke replied:
Thank you, corrected.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -502,5 +524,23 @@ def document_list(request):
+                         document.hazard_type,
+                         other_perils,
+                         document.hazard_unit,
+                         document.distribution_url,
```
Here you have duplicate distribution URL with line 539.

ErikKBethke replied:
Removed, thank you.

Vampouille commented on geonode/layers/views.py:
```diff
@@ -507,19 +508,66 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
+    response['Content-Disposition'] = 'attachment; filename=' + filename
     writer = csv.writer(response)
+    # Writes header row
+    writer.writerow(['Layer ID','Name',
```
Can you move "Name" header to next line so we have always one header label per line ?

Vampouille replied:
You also need to add one space to fix indentation

ErikKBethke replied:
Indentation and header line fixed.

Vampouille commented on geonode/layers/views.py:
```diff
@@ -507,19 +508,66 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
+                         layer.category,
+                         layer.abstract
+                         ])
+return response
```
Same apply here, return statement is not in function. You can check syntax of python files with following command : 
```bash
$ python -m py_compile geonode/layers/views.py
```
This will show you syntax errors like : 
```
  File "geonode/layers/views.py", line 570
    return response
SyntaxError: 'return' outside function
```

Vampouille replied:
To check all files :
```bash
$ python -m py_compile `find -name "*.py"`
```

ErikKBethke replied:
Indented return statement to include in function.

Thank you for the corrections and advice.

Vampouille commented on geonode/documents/views.py:
```diff
@@ -489,9 +490,30 @@ def document_metadata_detail(request, docid, template='documents/document_metada
     writer = csv.writer(response)
+    # Writes header row
+    writer.writerow(['Layer ID',
+                    'Title',
```
Wrong indentation here : you should indent `'Title'` and next lines to `'Layer ID'` so you need to add one space.

ErikKBethke replied:
Thank you, corrected.

Vampouille commented on geonode/layers/views.py:
```diff
@@ -507,19 +508,66 @@ def layer_change_poc(request, ids, template='layers/layer_change_poc.html'):
+                         layer.calculation_method_quality,
+                         layer.scientific_quality,
+                         settings.SITEURL + layer.get_absolute_url(),
+                         unicodedata.normalize('NFKD', layer.supplemental_information).encode('ascii','ignore'),
```
Missing space after `'ascii',` and before `'ignore'`

ErikKBethke replied:
Added space for consistency

ErikKBethke committed:
Fix indentation, header label, return statement

ErikKBethke committed:
Update indentation, variables

Update indentation within header writerow and return response. Fix variable names, proper writerow

ErikKBethke committed:
Add comma to writerow

Missed comma after name

ErikKBethke committed:
Update distribution_url

Distribution_url is deprecated (referenced here https://github.com/GFDRR/thinkhazard/issues/746) and already shows issues when trying to call documents/list function on new server. Change made to avoid this issue.

ErikKBethke commented:
Made a change to update deprecated field, but I noticed when running http://54.229.171.111/layers/list on the unmerged version of GeoNode that there are issues with unicode/ _proxy_ compatibility. Do you think this will continue to be an issue? @Vampouille

stufraser1 commented:
@Vampouille Please advise if this can be merged now - we need to use the udpated list views for some tasks to update data and maps. thanks

Vampouille commented on geonode/documents/views.py:
```diff
@@ -489,18 +490,58 @@ def document_metadata_detail(request, docid, template='documents/document_metada
+                     'License',
+                     'Restrictions',
+                     'Document URL',
+                     'Distribution URL',
```
Can you change header label to 'Detail URL' ?

Vampouille commented on geonode/documents/views.py:
```diff
@@ -531,7 +531,7 @@ def document_list(request):
                          other_perils,
                          document.hazard_unit,
-                         document.distribution_url,
+                         document.detail_url,
```
There is 18 labels on header for 19 fields of data. Can you add a header for detail_url field ?

Vampouille commented:
@ErikKBethke Change from `distribution_url` to `detail_url` only impact layer objects. So you don't need to make change, changes are already made on https://github.com/GFDRR/geonode/tree/hosting  (geonode 2.5). So can you revert https://github.com/GFDRR/geonode/pull/29/commits/41170777cd38a04e2b30265367aeeb5e83cd492a ?

This PR can be merged without last commit.

stufraser1 committed:
Update views.py

stufraser1 commented:
@Vampouille @ErikKBethke 
I reverted 4117077 as requested above (removed document.detail_url at line 534).
No header for document.detail_url had been added

Please review and merge if acceptable

Vampouille commented:
LGTM

stufraser1 commented:
@Vampouille I tried to run this, error still occurs at line 505: 'Document' object has no attribute 'distribution_url'
Checking http://www.geonode-gfdrrlab.org/api/documents/
I see a detail_url but no distribution_url - so in fact this should be changed in documents as well as layers.
Change committed

stufraser1 commented:
@Vampouille 
http://www.geonode-gfdrrlab.org/layers/list returns error: must be unicode, not __proxy__
at location: /usr/src/app/geonode/layers/views.py in layer_list, line 521
Are you able to assist in resolving this please?
I require this list view to work as soon as possible, to audit the data on the new geonode
Thank you

Vampouille commented:
I'm currently working on this feature but on new 2.6 upstream code. There is some "non ascii" characters in database, so we need to choose what to do with those characters. There is several choices : 

* convert all non ascii characters to corresponding ascii character : "Stéphane" will be exported as "Stephane"
* export in utf8, most of non ascii characters should be correctly exported.
* export in another charset of your choice. 

I would recommend second solution : export in UTF8 charset. @stufraser1 what is your preference ?

stufraser1 commented:
@Vampouille Happy to take your suggestion, if there are no issues with reading and editing UTF-8 in excel when the listview is exported to .csv - this will be the main use of it.

Vampouille commented:
I finish to implements CSV export for layers on geonode 2.6 with this commit :
https://github.com/camptocamp/labs_geonode/commit/b2a0a08ca7b35a302bdffbe8e31bd4560dd31052

Vampouille commented:
@stufraser1 About document csv export, document does not have `name` field so following code : 
https://github.com/GFDRR/geonode/pull/29/files#diff-bc991905f1c7cd04b63504af988bc9c1R529
is not working. Can I just remove `name` field ?

stufraser1 commented:
Yes please remove name field if it does not exist

On Fri, 12 May 2017, 10:21 Julien Acroute, <notifications@github.com> wrote:

> @stufraser1 <https://github.com/stufraser1> About document csv export,
> document does not have name field so following code :
>
> https://github.com/GFDRR/geonode/pull/29/files#diff-bc991905f1c7cd04b63504af988bc9c1R529
> is not working. Can I just remove name field ?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/GFDRR/geonode/pull/29#issuecomment-301028076>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AL724vcaiu9PtFuVuU3-i6LMVgC8Nk-aks5r5CSigaJpZM4M8y2b>
> .
>
-- 
Stuart Fraser, PhD
Disaster Risk Consultant
www.disaster-risk.uk
Skype: stu_fraser1
+44 (0)7455 048 044

Vampouille commented:
@stufraser1 Can you check if you have all needed informations in following csv files :

[documents_2017-05-12.xlsx](https://github.com/GFDRR/geonode/files/995988/documents_2017-05-12.xlsx)
[layers_2017-05-12.xlsx](https://github.com/GFDRR/geonode/files/995987/layers_2017-05-12.xlsx)

(On geonode, you will download CSV file, I had to convert csv to xlsx in order to add those files on github)

stufraser1 commented:
These files have everything we need. Thank you
